### ISSUE TITLE
Trouble with the interface for `transactionAsync`

### ISSUE THREAD

**Issue by cmoel** [ospfranco/react-native-quick-sqlite#92]
I'm trying to use the [Async Transactions](https://github.com/ospfranco/react-native-quick-sqlite#transactions) API and I'm running into a couple issues. It's possible I'm doing it wrong but it doesn't seem like it right now.

After opening the database, I'm purposefully raising an error, hoping to `catch` it on the JavaScript side:

```typescript
db.transactionAsync(tx =>
  tx
    .executeAsync("insert into dogs (name) values (?)", ["Sam"])
    .then(res =>
      tx.executeAsync("insert into people (dog_id, name) values (?, ?)", [
        res.insertId,
        "Kirsten",
      ]),
    )
    .then(() => tx.executeAsync(selectAll, []))
    .catch(e => console.log("ERROR", e)),
)
```

I haven't defined the `people` table yet so an error is raised.

```
SQL execution error: no such table: people
```

Which is what I expect. I'd like to `catch` this error rather than have a fatal React Native error:

<img width="483" alt="Screen Shot 2022-10-20 at 20 25 43" src="https://user-images.githubusercontent.com/272904/197085290-8c0e6c86-9a1e-4a16-acbe-bcb971c79efd.png">

Also, the database is locked after this failure and I'm not sure how to get around that. I assume it's because I was in the middle of a transaction when it failed?

<img width="483" alt="Screen Shot 2022-10-20 at 20 49 18" src="https://user-images.githubusercontent.com/272904/197086021-b23b3246-5aca-4095-83be-f625577229c9.png">

Lastly, the code snippet doesn't type check. I get this TypeScript error:

```
Type 'Promise<void | QueryResult>' is not assignable to type 'Promise<void>'.
  Type 'void | QueryResult' is not assignable to type 'void'.
    Type 'QueryResult' is not assignable to type 'void'.
```

To fix this, I'm adding a `Promise.resolve` call as the last step, which doesn't feel quite right:

```typescript
db.transactionAsync(tx =>
  tx
    .executeAsync("insert into dogs (name) values (?)", ["Sam"])
    .then(res =>
      tx.executeAsync("insert into people (dog_id, name) values (?, ?)", [
        res.insertId,
        "Kirsten",
      ]),
    )
    .then(() => tx.executeAsync(selectAll, []))
    .then(() => Promise.resolve())                // <----- Added this
    .catch(e => console.log("ERROR", e)),
)
```

I wonder if the typings should be updated to return either `Promise<QueryResult>` or `Promise<void>`?

**Comment by ospfranco** [comment]
hmm weird, you catch should work (I assume you are OK with a incorrect/partial insertion of data). Can you create a reproducible example?

**Comment by cmoel** [comment]
Yeah, I'll put together a repo with a minimal reproducible example. In this case, I wouldn't expect incorrect or partial insertion of data. I would expect the transaction to rollback, not commit.

**Comment by cmoel** [comment]
@ospfranco I made a minimal reproducible example at [cmoel/react-native-quick-sqlite-transaction-issues](https://github.com/cmoel/react-native-quick-sqlite-transaction-issues).

This example reproduces the `SQL execution error` that I would prefer to catch on the JavaScript side and the `Promise.resolve` call required to make the transaction functions type check.

**Comment by ospfranco** [comment]
you are catching the error though, how do you expect the code to detect an incorrect insertion?

**Comment by cmoel** [comment]
@ospfranco I would _like_ to catch the error, which is why I'm calling `catch` on the promise chains. They aren't being hit, though. I tried to clarify this in this commit: https://github.com/cmoel/react-native-quick-sqlite-transaction-issues/commit/195ff54d32fb528b1c3f84b896ec739f947c542e. Is there something else I'm missing or should be doing differently?

**Comment by ospfranco** [comment]
I might be wrong, but are you following the promise chain correctly? you are not returning the promise

![Screenshot 001196](https://user-images.githubusercontent.com/1634213/197514063-b2cd593e-a9dc-4162-a431-e05f995e1b4d.png)

**Comment by ospfranco** [comment]
Ok, I took a look. I know understand your last resolve, the types were incorrect, I changed the result of the fn param to Promise<any>.

As for the database locked, this one is a bit more misterious to me, it seems to me that since you catched the error you skipped the internal `rollback` statement and database is locked in the transaction.

**Comment by ospfranco** [comment]
Sorry, you are right, the error is not being properly catched, probably has to do with how I implemented the error throwing inside the async execute function. I will need to take a look later though.

**Comment by ospfranco** [comment]
I did a quick refactoring, seems to be working from a very quick testing, give it a try and let me know if it is working properly now (5.0.3)

**Comment by cmoel** [comment]
Cool, thanks! v5.0.3 now lets me catch the SQL error in the my app 🎉

One issue I've found, though, is that transactions aren't atomic: the first insert happens but isn't rolled back when the second insert fails. I should have time over the next few days to put together a TypeScript-only PR to address this.

**Comment by ospfranco** [comment]
you need to throw an error for the catch inside to rollback the entire transaction, that's what I meant: you are catching the error inside, the transaction it's not going to be rolled back for you

**Comment by ospfranco** [comment]
New transactions now have commit and rollback methods, they still internally catch any thrown error but there is a bit more control for anyone who would like to internally catch any error. Released on 6.0.0
